File: runtimes/swift/Sources/Twilic/Core/Session.swift
import Foundation

public enum UnknownReferencePolicy: Int, Sendable {
    case failFast = 0
    case statelessRetry = 1
}

public let unknownReferencePolicyFailFast = UnknownReferencePolicy.failFast
public let unknownReferencePolicyStatelessRetry = UnknownReferencePolicy.statelessRetry

public enum DictionaryFallback: Int, Sendable {
    case failFast = 0
    case statelessRetry = 1
}

public let dictionaryFallbackFailFast = DictionaryFallback.failFast
public let dictionaryFallbackStatelessRetry = DictionaryFallback.statelessRetry

func dictionaryFallbackFromByte(_ b: UInt8) -> (DictionaryFallback, Bool) {
    if b == 0 { return (.failFast, true) }
    if b == 1 { return (.statelessRetry, true) }
    return (.failFast, false)
}

public struct DictionaryProfile: Sendable {
    public var version: UInt64
    public var hash: UInt64
    public var expiresAt: UInt64
    public var fallback: DictionaryFallback
}

public struct SessionOptions: Sendable {
    public var maxBaseSnapshots: Int = 8
    public var enableStatePatch: Bool = true
    public var enableTemplateBatch: Bool = true
    public var enableTrainedDictionary: Bool = true
    public var unknownReferencePolicy: UnknownReferencePolicy = .failFast
}

public func defaultSessionOptions() -> SessionOptions { SessionOptions() }

public final class InternTable {
    private var byValue: [String: UInt64] = [:]
    private var byID: [String] = []

    func getID(_ value: String) -> (UInt64, Bool) {
        if let refID = byValue[value] { return (refID, true) }
        return (0, false)
    }

    func getValue(_ refID: UInt64) -> (String, Bool) {
        guard refID < UInt64(byID.count) else { return ("", false) }
        return (byID[Int(refID)], true)
    }

    @discardableResult
    func register(_ value: String) -> UInt64 {
        if let existing = byValue[value] { return existing }
        let refID = UInt64(byID.count)
        byID.append(value)
        byValue[value] = refID
        return refID
    }

    func clear() {
        byValue = [:]
        byID = []
    }

    var registeredStringsInOrder: [String] { byID }
}

func shapeKey(_ keys: [String]) -> String { keys.joined(separator: "\0") }

public final class ShapeTable {
    private var byKeys: [String: UInt64] = [:]
    private var byID: [UInt64: [String]] = [:]
    private var observations: [String: Int] = [:]
    private(set) var nextID: UInt64 = 0

    func getID(_ keys: [String]) -> (UInt64, Bool) {
        let sk = shapeKey(keys)
        if let refID = byKeys[sk] { return (refID, true) }
        return (0, false)
    }

    func getKeys(_ refID: UInt64) -> ([String], Bool) {
        if let keys = byID[refID] { return (keys, true) }
        return ([], false)
    }

    @discardableResult
    func register(_ keys: [String]) -> UInt64 {
        let sk = shapeKey(keys)
        if let existing = byKeys[sk] { return existing }
        let refID = nextID
        nextID += 1
        byID[refID] = keys
        byKeys[sk] = refID
        return refID
    }

    func registerWithID(_ shapeID: UInt64, _ keys: [String]) -> Bool {
        let sk = shapeKey(keys)
        if let existing = byID[shapeID] {
            return shapeKey(existing) == sk
        }
        if let existingID = byKeys[sk], existingID != shapeID { return false }
        byID[shapeID] = keys
        byKeys[sk] = shapeID
        if shapeID + 1 > nextID { nextID = shapeID + 1 }
        return true
    }

    func observe(_ keys: [String]) -> Int {
        let sk = shapeKey(keys)
        let count = (observations[sk] ?? 0) + 1
        observations[sk] = count
        return count
    }

    func clear() {
        byKeys = [:]
        byID = [:]
        observations = [:]
        nextID = 0
    }
}

public struct BaseSnapshotEntry {
    var id: UInt64
    var message: Message
}

public final class SessionState: @unchecked Sendable {
    public var options: SessionOptions = defaultSessionOptions()
    public let keyTable = InternTable()
    public let stringTable = InternTable()
    public let shapeTable = ShapeTable()
    public var encodeShapeObservations: [String: Int] = [:]
    public var baseSnapshots: [BaseSnapshotEntry] = []
    public var templates: [UInt64: TemplateDescriptor] = [:]
    public var templateColumns: [UInt64: [Column]] = [:]
    public var fieldEnums: [String: [String]] = [:]
    public var dictionaries: [UInt64: Data] = [:]
    public var dictionaryProfiles: [UInt64: DictionaryProfile] = [:]
    public var schemas: [UInt64: Schema] = [:]
    public var lastSchemaID: UInt64?
    public var previousMessage: Message?
    public var previousMessageSize: Int?
    public var nextBaseID: UInt64 = 0
    public var nextTemplateID: UInt64 = 0
    public var nextDictionaryID: UInt64 = 0
}

public func newSessionState() -> SessionState { SessionState() }

public func newSessionStateWithOptions(_ options: SessionOptions) -> SessionState {
    let s = SessionState()
    s.options = options
    return s
}

func registerBaseSnapshot(_ state: SessionState, _ baseID: UInt64, _ message: Message) {
    var filtered = state.baseSnapshots.filter { $0.id != baseID }
    filtered.append(BaseSnapshotEntry(id: baseID, message: message.clone()))
    while filtered.count > state.options.maxBaseSnapshots {
        filtered.removeFirst()
    }
    state.baseSnapshots = filtered
}

func allocateBaseID(_ state: SessionState) -> UInt64 {
    let refID = state.nextBaseID
    state.nextBaseID += 1
    return refID
}

func allocateTemplateID(_ state: SessionState) -> UInt64 {
    let refID = state.nextTemplateID
    state.nextTemplateID += 1
    return refID
}

func allocateDictionaryID(_ state: SessionState) -> UInt64 {
    let refID = state.nextDictionaryID
    state.nextDictionaryID += 1
    return refID
}

func getBaseSnapshot(_ state: SessionState, _ baseID: UInt64) -> (Message?, Bool) {
    for entry in state.baseSnapshots where entry.id == baseID {
        return (entry.message.clone(), true)
    }
    return (nil, false)
}

func resetTables(_ state: SessionState) {
    state.keyTable.clear()
    state.stringTable.clear()
    state.shapeTable.clear()
    state.encodeShapeObservations = [:]
    state.fieldEnums = [:]
}

func resetState(_ state: SessionState) {
    resetTables(state)
    state.baseSnapshots = []
    state.templates = [:]
    state.templateColumns = [:]
    state.dictionaries = [:]
    state.dictionaryProfiles = [:]
    state.schemas = [:]
    state.lastSchemaID = nil
    state.previousMessage = nil
    state.previousMessageSize = nil
    state.nextBaseID = 0
    state.nextTemplateID = 0
    state.nextDictionaryID = 0
}

File: runtimes/r/R/session.R
UnknownReferencePolicyFailFast <- 0L
UnknownReferencePolicyStatelessRetry <- 1L
DictionaryFallbackFailFast <- 0L
DictionaryFallbackStatelessRetry <- 1L

dictionary_fallback_from_byte <- function(b) {
  if (b == 0L) return(list(DictionaryFallbackFailFast, TRUE))
  if (b == 1L) return(list(DictionaryFallbackStatelessRetry, TRUE))
  list(DictionaryFallbackFailFast, FALSE)
}

default_session_options <- function() {
  list(
    max_base_snapshots = 8L,
    enable_state_patch = TRUE,
    enable_template_batch = TRUE,
    enable_trained_dictionary = TRUE,
    unknown_reference_policy = UnknownReferencePolicyFailFast
  )
}

new_intern_table <- function() {
  env <- new.env(parent = emptyenv())
  env$by_value <- list()
  env$by_id <- list()
  env
}

intern_table_get_id <- function(table, value) {
  id <- table$by_value[[value]]
  if (!is.null(id)) return(list(id, TRUE))
  list(0L, FALSE)
}

intern_table_get_value <- function(table, ref_id) {
  ref_id <- as.integer(ref_id) + 1L
  if (ref_id > length(table$by_id)) return(list("", FALSE))
  list(table$by_id[[ref_id]], TRUE)
}

intern_table_register <- function(table, value) {
  existing <- table$by_value[[value]]
  if (!is.null(existing)) return(existing)
  ref_id <- length(table$by_id)
  table$by_id[[ref_id + 1L]] <- value
  table$by_value[[value]] <- ref_id
  ref_id
}

intern_table_clear <- function(table) {
  table$by_value <- list()
  table$by_id <- list()
  invisible(NULL)
}

shape_key <- function(keys) {
  if (!length(keys)) return("")
  sep <- rawToChar(as.raw(0L))
  Reduce(function(a, b) paste0(a, sep, b), keys)
}

new_null_byte_map <- function() new.env(parent = emptyenv(), hash = TRUE)

shape_table_get_id <- function(table, keys) {
  sk <- shape_key(keys)
  id <- table$by_keys[[sk]]
  if (!is.null(id)) return(list(id, TRUE))
  list(0L, FALSE)
}

shape_table_get_keys <- function(table, ref_id) {
  keys <- table$by_id[[as.character(ref_id)]]
  if (is.null(keys)) return(list(NULL, FALSE))
  list(keys, TRUE)
}

new_shape_table <- function() {
  list(
    by_keys = new_null_byte_map(),
    by_id = new_null_byte_map(),
    observations = new_null_byte_map(),
    next_id = 0L
  )
}

shape_table_register <- function(table, keys) {
  sk <- shape_key(keys)
  existing <- table$by_keys[[sk]]
  if (!is.null(existing)) return(existing)
  ref_id <- table$next_id
  table$next_id <- table$next_id + 1L
  table$by_id[[as.character(ref_id)]] <- keys
  table$by_keys[[sk]] <- ref_id
  ref_id
}

shape_table_register_with_id <- function(table, shape_id, keys) {
  sk <- shape_key(keys)
  existing <- table$by_id[[as.character(shape_id)]]
  if (!is.null(existing) && shape_key(existing) != sk) return(FALSE)
  existing_id <- table$by_keys[[sk]]
  if (!is.null(existing_id) && existing_id != shape_id) return(FALSE)
  table$by_id[[as.character(shape_id)]] <- keys
  table$by_keys[[sk]] <- shape_id
  if (shape_id + 1L > table$next_id) table$next_id <- shape_id + 1L
  TRUE
}

shape_register_with_id <- shape_table_register_with_id

shape_table_observe <- function(table, keys) {
  sk <- shape_key(keys)
  table$observations[[sk]] <- (table$observations[[sk]] %||% 0L) + 1L
  table$observations[[sk]]
}

shape_table_clear <- function(table) {
  table$by_keys <- new_null_byte_map()
  table$by_id <- new_null_byte_map()
  table$observations <- new_null_byte_map()
  table$next_id <- 0L
  table
}

new_session_state_with_options <- function(options) {
  new_session_state(options)
}

new_session_state <- function(options = NULL) {
  opts <- options %||% default_session_options()
  state <- new.env(parent = emptyenv())
  state$options <- opts
  state$key_table <- new_intern_table()
  state$string_table <- new_intern_table()
  state$shape_table <- new_shape_table()
  state$encode_shape_observations <- new_null_byte_map()
  state$base_snapshots <- list()
  state$templates <- new_null_byte_map()
  state$template_columns <- new_null_byte_map()
  state$field_enums <- new_null_byte_map()
  state$dictionaries <- new_null_byte_map()
  state$dictionary_profiles <- new_null_byte_map()
  state$schemas <- new_null_byte_map()
  state$last_schema_id <- NULL
  state$previous_message <- NULL
  state$previous_message_size <- NULL
  state$next_base_id <- 0L
  state$next_template_id <- 0L
  state$next_dictionary_id <- 0L
  state
}

register_base_snapshot <- function(state, base_id, message) {
  filtered <- Filter(function(e) e$id != base_id, state$base_snapshots)
  filtered[[length(filtered) + 1L]] <- list(id = base_id, message = message_clone(message))
  while (length(filtered) > state$options$max_base_snapshots) filtered <- filtered[-1]
  state$base_snapshots <- filtered
  state
}

allocate_base_id <- function(state) {
  id <- state$next_base_id
  state$next_base_id <- state$next_base_id + 1L
  id
}

allocate_template_id <- function(state) {
  id <- state$next_template_id
  state$next_template_id <- state$next_template_id + 1L
  id
}

allocate_dictionary_id <- function(state) {
  id <- state$next_dictionary_id
  state$next_dictionary_id <- state$next_dictionary_id + 1L
  id
}

get_base_snapshot <- function(state, base_id) {
  for (e in state$base_snapshots) {
    if (e$id == base_id) return(list(message_clone(e$message), TRUE))
  }
  list(NULL, FALSE)
}

reset_tables <- function(state) {
  intern_table_clear(state$key_table)
  intern_table_clear(state$string_table)
  state$shape_table <- shape_table_clear(state$shape_table)
  state$encode_shape_observations <- new_null_byte_map()
  state$field_enums <- new_null_byte_map()
  state
}

reset_state <- function(state) {
  state <- reset_tables(state)
  state$base_snapshots <- list()
  state$templates <- new_null_byte_map()
  state$template_columns <- new_null_byte_map()
  state$dictionaries <- new_null_byte_map()
  state$dictionary_profiles <- new_null_byte_map()
  state$schemas <- new_null_byte_map()
  state$last_schema_id <- NULL
  state$previous_message <- NULL
  state$previous_message_size <- NULL
  state$next_base_id <- 0L
  state$next_template_id <- 0L
  state$next_dictionary_id <- 0L
  state
}
